### ISSUE TITLE
Remove spaces from %HOSTNAME%

### ISSUE THREAD

**Issue by m0bYczE** [rsyslog/rsyslog#3717]
Hello, I need help with replacing spaces with "_" in %HOSTNAME%, beacuse when in %HOSTNAME% is space, the folder which I it creates have in name only first word from %HOSTNAME%. Here is my code:
`template (name="Mikrotik" type="string" string="/var/log/mikrotik/%HOSTNAME%/%$YEAR%-%$MONTH%-%$DAY%.log")

if $fromhost-ip startswith '192.168.' then {
	action(type="omfile" dynaFile="Mikrotik")
	stop
}`

Thank you.

**Comment by rgerhards** [comment]
I wonder how spaces can be present in HOSTNAME at all. What is your rsyslog version? Please add

*.* /var/log/msgdebug.log;RSYSLOG_DebugFormat

to the **top** of rsyslog.conf. That file contains multiple lines per message.  After the rsyslog run, search for a message with space in hostname and post all lines that belong to it. Thx.

**Comment by m0bYczE** [comment]
Version is: 8.24.0 and the debug:
`Debug line with all properties:
FROMHOST: '**.**.company.com', fromhost-ip: '**.**.**.**', HOSTNAME: 'Bytovka', PRI: 46,
syslogtag 'Tovarni', programname: 'Tovarni', APP-NAME: 'Tovarni', PROCID: '-', MSGID: '-',
TIMESTAMP: 'Jun 23 12:13:10', STRUCTURED-DATA: '-',
msg: ' X fttb sys: user system logged out from **.**.**.** via api'
escaped msg: ' X fttb sys: user system logged out from **.**.**.** via api'
inputname: imudp rawmsg: '<46>Jun 23 12:13:10 Bytovka Tovarni X fttb sys: user system logged out from **.**.**.** via api'
$!:
$.:
$/:`

**Comment by rgerhards** [comment]
Mmhhh looks valid. Hostname is "Bytovka" but some of the fields are in wrong order.

**Comment by rgerhards** [comment]
Or do you mean the system hostname really contains spaces? That is against all standards and not permitted.

**Comment by davidelang** [comment]
I suspect that he is intending the hostnme to be "Bytovka Tovarni X fttb" or 
something like that.

A space is not a valid character in a hostname or dns name. If you try to use a 
hostname with a space, all sorts of odd things are going to happen.

I don't see rsyslog trying to guess if a space indictes the end of a hostname or 
not when receiving messages. I guess we could log a warning if we start up on a 
system that has an invalid hostname (either invalid characters, or a hostname 
that's too long). But I don't see that as a high priority, and I don't see it 
helping the original poster.

David Lang

**Comment by m0bYczE** [comment]
Yes, the hostname is Bytovka Tovarni X fttb. Ok, i have to change hostname in other devices. Thank you.

**Comment by rgerhards** [comment]
> I guess we could log a warning if we start up on a system that has an invalid hostname (either invalid characters, or a hostname that's too long).

The problem here is how do you know if this is a "hostname-space" vs a regular one? You could try to guess that the tag stops with a ':' and search for it, then go backwards to the hostname. But tags not always stop with colon. They explicitly do not need to do in RFC5424. So very very fishy...

**Comment by davidelang** [comment]
I as thinking in terms of rsyslog checking the hostname of the system on the 
sender side.

David Lang

**Comment by rgerhards** [comment]
That's a good idea, but obviously does not work for other senders.

**Comment by davidelang** [comment]
On Sun, 23 Jun 2019, Rainer Gerhards wrote:

> That's a good idea, but obviously does not work for other senders.

Right, we can't know what's happening on messages we receive, but we can make 
sure we aren't part of the problem when we send.

I'm thinking of something that is checked when $$hostname is populated

David Lang

**Comment by skupkosk** [comment]
Just for a note.
Even the '_' character in question is not allowed by standard and rsyslog
does not handle it.
https://en.wikipedia.org/wiki/Hostname#Restrictions_on_valid_hostnames

**Comment by rgerhards** [comment]
> Right, we can't know what's happening on messages we receive, but we can make sure we aren't part of the problem when we send. I'm thinking of something that is checked when $$hostname is populated

This check is already done. With the obsolete legacy config system, we simply cannot detect a hard error but issue:
```
$localhostname hostname with blanks
```
results in:
```
rsyslogd: error: extra characters in config line ignored: 'with blanks' [v8.1907.0]
```

In current style `global(localhostname="hostname with blanks")` we emit a hard error and terminate startup.

The only way to get a hostname with spaces into the system is configuring it to use an invalid hostname. However, I do not know how to do this. This is what I get when I try:
```
$ hostname "my invalid name"
hostname: the specified hostname is invalid
```
In conclusion, I don't think there is anything we can do from the rsyslog side.
